util/src/main/java/io/kubernetes/client/Discovery.java

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package io.kubernetes.client;
+
+import com.google.common.base.Objects;
+import io.kubernetes.client.openapi.ApiClient;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.ApiResponse;
+import io.kubernetes.client.openapi.Configuration;
+import io.kubernetes.client.openapi.models.V1APIGroup;
+import io.kubernetes.client.openapi.models.V1APIGroupList;
+import io.kubernetes.client.openapi.models.V1APIResourceList;
+import io.kubernetes.client.openapi.models.V1APIVersions;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import okhttp3.Call;
+
+public class Discovery {
+
+  private final ApiClient apiClient;
+
+  public Discovery() {
+    this(Configuration.getDefaultApiClient());
+  }
+
+  public Discovery(ApiClient apiClient) {
+    this.apiClient = apiClient;
+  }
+
+  public Set<APIResource> findAll() throws ApiException {
+    Set<APIResource> allResources = new HashSet<>();
+    for (String version : legacyCoreApi().getVersions()) {
+      allResources.addAll(findAll("", Arrays.asList(version), version, "/api/" + version));
+    }
+    for (V1APIGroup group : groupDiscovery("/apis").getGroups()) {
+      allResources.addAll(
+          findAll(
+              group.getName(),
+              group.getVersions().stream().map(v -> v.getVersion()).collect(Collectors.toList()),
+              group.getPreferredVersion().getVersion()));
+    }
+    return allResources;
+  }
+
+  public Set<APIResource> findAll(String group, List<String> versions, String preferredVersion)
+      throws ApiException {
+    return findAll(group, versions, preferredVersion, "/apis/" + group + "/" + preferredVersion);
+  }
+
+  public Set<APIResource> findAll(
+      String group, List<String> versions, String preferredVersion, String path)
+      throws ApiException {
+    V1APIResourceList resourceList = resourceDiscovery(path);
+    return sort(group, versions, preferredVersion, resourceList);
+  }
+
+  private Set<APIResource> sort(
```

Comment:
it feels like `sort` isn't quite the right name for this method as it doesn't seem to be sorting anything. Perhaps a more descriptive name?

Reply:
renamed to `groupResourcesByName` and added descriptive comments